### PR TITLE
Add parameter values for ftp_proxy and http_proxy.

### PR DESCRIPTION
**Pull request by Joshua-Snapp**
Some servers are not allowed to directly access ftp and http addresses.
Let's provide the option for someone to be able to set the ftp_proxy or
http_proxy values in the cpan Config.pm.

### CODE REVIEW

File: templates/cpan.conf.erb [hunk 1]
```diff
@@ -15,14 +15,14 @@ $CPAN::Config = {
   'curl' => q[/usr/bin/curl],
   'ftp' => q[/usr/bin/ftp],
   'ftp_passive' => q[1],
-  'ftp_proxy' => q[],
+  'ftp_proxy' => q[<%=@ftp_proxy%>],
```
Comment (salimane) [inline templates/cpan.conf.erb]: please check ``@ftp_proxy`` is not undefined before adding this line in the config

File: templates/cpan.conf.erb [hunk 2]
```diff
@@ -15,14 +15,14 @@ $CPAN::Config = {
...
   'getcwd' => q[cwd],
   'gpg' => q[/usr/bin/gpg],
   'gzip' => q[/bin/gzip],
   'halt_on_failure' => q[0],
   'histfile' => q[/root/.cpan/histfile],
   'histsize' => q[100],
-  'http_proxy' => q[],
+  'http_proxy' => q[<%=@http_proxy%>],
```
Comment (salimane) [inline templates/cpan.conf.erb]: please check ``@ http_proxy `` is not undefined before adding this line in the config

**Comment by Joshua-Snapp** [comment]
I haven't done much with testing yet as I have been learning Puppet. I took a look at your tests and tried to copy what you had done for local_lib. Let me know if there's more that I need to do with that.

Do you have a link to info that describes the need to test whether a parameter is not undef before setting that value in a template? Or perhaps you could take a moment to describe the reason behind that. I just want to understand better why that might be the right thing to do, instead of just throwing the value in the template the way I did originally.

Thanks for your time, and your module!

**Comment by salimane** [comment]
@Joshua-Snapp you set the default value of ftp_proxy to undef, so it means that in the template when puppet runs and try to provision the file the value would be nil or empty. is that a right configuration for cpan ? It may not be. Also it's generally good not to set configs that are not needed.

For the tests, you should adding your rspec examples in [cpan_spec.rb](https://github.com/meltwater/puppet-cpan/blob/master/spec/classes/cpan_spec.rb) . Check [here](https://github.com/meltwater/puppet-marathon/blob/master/spec/classes/init_spec.rb#L128) for some examples on how to do this.

Thanks

**Comment by Joshua-Snapp** [comment]
@salimane The default Config.pm for a fresh install of cpan sets ftp_proxy and http_proxy to empty values. The existing [cpan.conf.erb template](https://github.com/meltwater/puppet-cpan/blob/master/templates/cpan.conf.erb#L18) file in this repo sets empty values for both ftp_proxy and http_proxy. So, yes, I believe that to be a right configuration for cpan. They are good default values, since most machines don't require a proxy to access cpan mirrors.

My organization has some machines that _do_ require a proxy to access external ftp and http addresses. I need to set ftp_proxy and http_proxy in the Config.pm file. My fork of this repo provides that flexibility.

As for the testing, I appreciate the link you provided as an example of what that code looks like. However, I don't see where testing is being done in [cpan_spec.rb](https://github.com/meltwater/puppet-cpan/blob/master/spec/classes/cpan_spec.rb) for the other parameters that define values in the cpan.conf.erb template file. Please point me to the tests for installdirs and config_hash, and I should be able to figure out how to add tests for setting ftp_proxy and http_proxy.

File: manifests/init.pp [hunk 3]
```diff
@@ -45,6 +53,8 @@
   validate_bool($local_lib)
   validate_string($config_template)
   validate_string($package_ensure)
+  validate_string($ftp_proxy)
+  validate_string($http_proxy)
```
Comment (alexjfisher) [inline manifests/init.pp]: If the defaults for the new parameters are `undef`, these validations will fail.

File: templates/cpan.conf.erb [hunk 4]
```diff
@@ -15,14 +15,14 @@ $CPAN::Config = {
...
-  'ftp_proxy' => q[],
+  'ftp_proxy' => q[<% if @ftp_proxy != "NONE" %><%= @ftp_proxy %><% end %>],
   'getcwd' => q[cwd],
   'gpg' => q[/usr/bin/gpg],
   'gzip' => q[/bin/gzip],
   'halt_on_failure' => q[0],
   'histfile' => q[/root/.cpan/histfile],
   'histsize' => q[100],
-  'http_proxy' => q[],
+  'http_proxy' => q[<% if @http_proxy != "NONE" %><%= @http_proxy %><% end %>],
```
Comment (alexjfisher) [inline templates/cpan.conf.erb]: parameters default to `undef`, not the string `'NONE'`

**Comment by alexjfisher** [comment]
I'm interested in this feature too.  I can open a new PR with an extra commit that adds tests and fixes a couple of issues.

**Comment by salimane** [comment]
@alexjfisher please do 👍

File: manifests/init.pp (hunk 3 above)
Comment (alexjfisher) [inline manifests/init.pp]: I take that back.
> Note: validate_string(undef) will not fail in this version of the functions API.
https://github.com/puppetlabs/puppetlabs-stdlib#validate_string

**Comment by salimane** [comment]
closed in favor of #39

**Pull request closed without merging**